File: scripts/lib/sparse_merkle.ts
import { SPARSE_MERKLE_DEPTH, ZERO_FIELD } from "./constants.js";
import { fieldToBytesBE, fieldToHex, bytesBEToField } from "./field.js";
import { initPoseidon2, poseidon2HashFixed } from "./poseidon2.js";

export interface SparseNonMembershipProof {
  key: bigint;
  root: bigint;
  path: bigint[];
  pathIndices: number[];
}

/** Matches contracts/asp_nonmembership sparse_merkle.rs */
export class SparseMerkleTree {
  private leafMap = new Map<number, bigint>();
  private nodeCache = new Map<string, bigint>();
  private defaults: bigint[] = [];
  private root: bigint = ZERO_FIELD;

  constructor(private depth: number = SPARSE_MERKLE_DEPTH) {}

  async init(): Promise<void> {
    await initPoseidon2();
    this.defaults = new Array(this.depth + 1);
    this.defaults[this.depth] = ZERO_FIELD;
    for (let d = this.depth - 1; d >= 0; d--) {
      const child = this.defaults[d + 1]!;
      this.defaults[d] = await poseidon2HashFixed([child, child]);
    }
    this.root = this.defaults[0]!;
  }

  getDepth(): number {
    return this.depth;
  }

  /** Last 4 bytes BE masked — same as sparse_merkle::key_to_index */
  private keyToIndex(key: bigint): number {
    const bytes = fieldToBytesBE(key);
    const low =
      (bytes[28]! << 24) |
      (bytes[29]! << 16) |
      (bytes[30]! << 8) |
      bytes[31]!;
    return low & ((1 << this.depth) - 1);
  }

  /** poseidon2_hash_2(pattern, poseidon2_hash_1([1u8; 32])) — Rust [1u8; 32] is 32 bytes of 1 */
  private async patternLeaf(key: bigint): Promise<bigint> {
    const oneBytes = new Uint8Array(32).fill(1);
    const hashedOne = await poseidon2HashFixed([bytesBEToField(oneBytes)]);
    return poseidon2HashFixed([key, hashedOne]);
  }

  private async hashAtLevel(level: number, index: number): Promise<bigint> {
    if (level === this.depth) {
      return this.leafMap.get(index) ?? this.defaults[this.depth]!;
    }
    const cached = this.nodeCache.get(`${level}-${index}`);
    if (cached !== undefined) {
      return cached;
    }
    return this.defaults[level]!;
  }

  async insert(key: bigint): Promise<void> {
    const leaf = await this.patternLeaf(key);
    const index = this.keyToIndex(key);
    this.leafMap.set(index, leaf);

    let node = leaf;
    let idx = index;
    for (let level = this.depth; level >= 1; level--) {
      const siblingIndex = idx ^ 1;
      const sibling = await this.hashAtLevel(level, siblingIndex);
      node =
        (idx & 1) === 1
          ? await poseidon2HashFixed([sibling, node])
          : await poseidon2HashFixed([node, sibling]);
      idx >>= 1;
      this.nodeCache.set(`${level - 1}-${idx}`, node);
    }
    this.root = node;
  }

  has(key: bigint): boolean {
    return this.leafMap.has(this.keyToIndex(key));
  }

  getRoot(): bigint {
    return this.root;
  }

  async getNonMembershipProof(key: bigint): Promise<SparseNonMembershipProof> {
    if (this.has(key)) {
      throw new Error("Key is in tree — cannot produce non-membership proof");
    }

    const index = this.keyToIndex(key);
    const path: bigint[] = [];
    const pathIndices: number[] = [];

    for (let level = this.depth; level >= 1; level--) {
      const posAtLevel = index >> (this.depth - level);
      const siblingPos = posAtLevel ^ 1;
      path.push(await this.hashAtLevel(level, siblingPos));
      pathIndices.push(posAtLevel & 1);
    }

    return {
      key,
      root: this.root,
      path,
      pathIndices,
    };
  }

  async verifyNonMembership(proof: SparseNonMembershipProof): Promise<boolean> {
    if (this.has(proof.key)) {
      return false;
    }

    let current = this.defaults[this.depth]!;

    for (let i = 0; i < proof.path.length; i++) {
      const sibling = proof.path[i]!;
      const isRight = proof.pathIndices[i]!;
      current =
        isRight === 1
          ? await poseidon2HashFixed([sibling, current])
          : await poseidon2HashFixed([current, sibling]);
    }

    return current === proof.root;
  }

  getLeaves(): { key: string; leaf: string }[] {
    return [...this.leafMap.entries()].map(([index, leaf]) => ({
      key: String(index),
      leaf: fieldToHex(leaf),
    }));
  }
}

/** Fraud billing pattern hash per requirements */
export async function billingPatternHash(
  icdCategoryHash: bigint,
  amountBucketHash: bigint,
  providerTypeHash: bigint,
): Promise<bigint> {
  await initPoseidon2();
  return poseidon2HashFixed([
    icdCategoryHash,
    amountBucketHash,
    providerTypeHash,
  ]);
}

File: contracts/asp_nonmembership/src/sparse_merkle.rs
use soroban_sdk::{BytesN, Env, Vec};
use zklaim_common::{poseidon2_hash_1, poseidon2_hash_2, zero_field};

use crate::{CacheKey, DataKey};

pub const SPARSE_DEPTH: u32 = 16;

pub fn key_to_index(key: &BytesN<32>) -> u32 {
    let arr = key.to_array();
    let low = u32::from_be_bytes([arr[28], arr[29], arr[30], arr[31]]);
    low & ((1u32 << SPARSE_DEPTH) - 1)
}

pub fn pattern_leaf(env: &Env, pattern: &BytesN<32>) -> BytesN<32> {
    poseidon2_hash_2(env, pattern, &poseidon2_hash_1(env, &BytesN::from_array(env, &[1u8; 32])))
}

pub fn default_hash(env: &Env, level: u32) -> BytesN<32> {
    if level == SPARSE_DEPTH {
        return zero_field(env);
    }
    let key = CacheKey::Default(level);
    if let Some(v) = env.storage().instance().get(&key) {
        return v;
    }
    let child = default_hash(env, level + 1);
    let node = poseidon2_hash_2(env, &child, &child);
    env.storage().instance().set(&key, &node);
    node
}

pub fn empty_root(env: &Env) -> BytesN<32> {
    default_hash(env, 0)
}

fn hash_at_level(env: &Env, level: u32, index: u32) -> BytesN<32> {
    if level == SPARSE_DEPTH {
        let key = DataKey::Leaf(index);
        if env.storage().persistent().has(&key) {
            return env.storage().persistent().get(&key).unwrap();
        }
        return default_hash(env, SPARSE_DEPTH);
    }
    env.storage()
        .instance()
        .get(&CacheKey::Node(level, index))
        .unwrap_or_else(|| default_hash(env, level))
}

fn cache_node(env: &Env, level: u32, index: u32, value: &BytesN<32>) {
    env.storage()
        .instance()
        .set(&CacheKey::Node(level, index), value);
}

pub fn insert_leaf(env: &Env, index: u32, leaf: &BytesN<32>) -> BytesN<32> {
    env.storage()
        .persistent()
        .set(&DataKey::Leaf(index), leaf);

    let mut node = leaf.clone();
    let mut idx = index;
    for level in (1..=SPARSE_DEPTH).rev() {
        let sibling_index = idx ^ 1;
        let sibling = hash_at_level(env, level, sibling_index);
        node = if idx & 1 == 1 {
            poseidon2_hash_2(env, &sibling, &node)
        } else {
            poseidon2_hash_2(env, &node, &sibling)
        };
        idx >>= 1;
        cache_node(env, level - 1, idx, &node);
    }
    node
}

pub fn get_non_membership_path(
    env: &Env,
    key: &BytesN<32>,
) -> (Vec<BytesN<32>>, Vec<u32>) {
    let index = key_to_index(key);
    let mut path = Vec::new(env);
    let mut path_indices = Vec::new(env);

    for level in (1..=SPARSE_DEPTH).rev() {
        let pos_at_level = index >> (SPARSE_DEPTH - level);
        let sibling_pos = pos_at_level ^ 1;
        path.push_back(hash_at_level(env, level, sibling_pos));
        path_indices.push_back(pos_at_level & 1);
    }

    (path, path_indices)
}

pub fn verify_non_membership(
    env: &Env,
    _key: &BytesN<32>,
    path: &Vec<BytesN<32>>,
    path_indices: &Vec<u32>,
    root: &BytesN<32>,
) -> bool {
    if path.len() as u32 != SPARSE_DEPTH || path_indices.len() as u32 != SPARSE_DEPTH {
        return false;
    }

    let mut current = default_hash(env, SPARSE_DEPTH);
    for i in 0..SPARSE_DEPTH {
        let sibling = path.get(i).unwrap();
        let is_right = path_indices.get(i).unwrap();
        current = if is_right == 1 {
            poseidon2_hash_2(env, &sibling, &current)
        } else {
            poseidon2_hash_2(env, &current, &sibling)
        };
    }

    current == *root
}
